crates/core/src/lib.rs

```diff
@@ -16,17 +16,605 @@ pub mod network;
+            nodes_config,
+        }
+    }
+}
+
+pub struct CoreBuilderWithNodesConfiguration {
+    metadata_store_client: MetadataStoreClient,
+    node_name: String,
+    force_node_id: Option<NodeId>,
+    advertise_address: AdvertisedAddress,
+    roles: EnumSet<Role>,
+    nodes_config: NodesConfiguration,
+}
+
+impl CoreBuilderWithNodesConfiguration {
+    pub fn with_networking<T: TransportConnect>(
+        self,
+        networking: Networking<T>,
+        metadata_builder: MetadataBuilder,
+    ) -> CoreBuilderWithNetworking<T> {
+        let mut router_builder = MessageRouterBuilder::default();
+        let metadata = metadata_builder.to_metadata();
+        let metadata_manager = MetadataManager::new(
+            metadata_builder,
+            networking.clone(),
+            self.metadata_store_client.clone(),
+        );
+        metadata_manager.register_in_message_router(&mut router_builder);
+
+        CoreBuilderWithNetworking {
+            node_name: self.node_name,
+            force_node_id: self.force_node_id,
+            advertise_address: self.advertise_address,
+            roles: self.roles,
+            nodes_config: self.nodes_config,
+            metadata_manager,
+            metadata_store_client: self.metadata_store_client,
+            networking,
+            metadata,
+        }
+    }
+
+    pub fn with_grpc_networking(
+        self,
+        networking_options: NetworkingOptions,
+    ) -> CoreBuilderWithNetworking<GrpcConnector> {
+        let metadata_builder = MetadataBuilder::default();
+        let networking = Networking::new(metadata_builder.to_metadata(), networking_options);
+
+        self.with_networking(networking, metadata_builder)
+    }
+}
+
+pub struct CoreBuilderWithNetworking<T> {
+    metadata_store_client: MetadataStoreClient,
+    node_name: String,
+    force_node_id: Option<NodeId>,
+    advertise_address: AdvertisedAddress,
+    roles: EnumSet<Role>,
+    nodes_config: NodesConfiguration,
+    metadata_manager: MetadataManager<T>,
+    metadata: Metadata,
+    networking: Networking<T>,
+}
+
+impl<T> CoreBuilderWithNetworking<T> {
+    pub fn with_tc(self, tc: TaskCenter) -> CoreBuilderFinal<T> {
+        CoreBuilderFinal {
+            tc,
+            node_name: self.node_name,
+            force_node_id: self.force_node_id,
+            advertise_address: self.advertise_address,
+            roles: self.roles,
+            nodes_config: self.nodes_config,
+            metadata_manager: self.metadata_manager,
+            metadata_store_client: self.metadata_store_client,
+            networking: self.networking,
+            metadata: self.metadata,
+            provider_kind: ProviderKind::Local,
+            allow_bootstrap: false,
+            partition_table: PartitionTable::with_equally_sized_partitions(Version::MIN, 10),
+            replication_strategy: ReplicationStrategy::OnAllNodes,
+            router_builder: MessageRouterBuilder::default(),
+        }
+    }
```

Comment:
The problem with the overall builder design (this was more so before your changes, so it's not feedback on your particular changes) is that it every starting point picks some defaults for some other components and we don't necessarily provide ergonomic way to edit them.

We can provide some shortcut functions that construct the builder with some common use-case defaults. `Core::builder_for_tests()`, and a default empty builder `Core::builder()`. Then one can use its internal sub-builders for an ergonomic API to configure it (especially in testing). `builder.nodes_config.add_node(...)`, etc.

This is just food for thought.